addons/binding/org.openhab.binding.smartthings/src/main/java/org/openhab/binding/smartthings/internal/SmartthingsHandlerFactory.java

```diff
@@ -0,0 +1,185 @@
+import org.openhab.binding.smartthings.handler.SmartthingsBridgeHandler;
+import org.openhab.binding.smartthings.handler.SmartthingsThingHandler;
+import org.osgi.framework.ServiceRegistration;
+import org.osgi.service.event.Event;
+import org.osgi.service.event.EventHandler;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.gson.Gson;
+
+/**
+ * The {@link SmartthingsHandlerFactory} is responsible for creating things and thing
+ * handlers.
+ *
+ * @author Bob Raker - Initial contribution
+ */
+public class SmartthingsHandlerFactory extends BaseThingHandlerFactory implements EventHandler {
+
+    private Logger logger = LoggerFactory.getLogger(SmartthingsHandlerFactory.class);
+    private Map<ThingUID, ServiceRegistration<?>> discoveryServiceRegs = new HashMap<>();
+
+    private SmartthingsBridgeHandler bridge;
+    private ChannelTypeRegistry channelTypeRegistry;
+    private Gson gson;
+    private List<SmartthingsThingHandler> thingHandlers = new ArrayList<SmartthingsThingHandler>();
+
+    @Override
+    public boolean supportsThingType(ThingTypeUID thingTypeUID) {
+        return THING_TYPE_SMARTTHINGS.equals(thingTypeUID) || SUPPORTED_THING_TYPES_UIDS.contains(thingTypeUID);
+    }
+
+    public SmartthingsHandlerFactory() {
+        // Get a Gson instance
+        gson = new Gson();
+    }
+
+    @Override
+    protected ThingHandler createHandler(Thing thing) {
+        ThingTypeUID thingTypeUID = thing.getThingTypeUID();
+        logger.debug("SmartthingsHandlerFactory is now processing ThingTypeUID {}", thingTypeUID.getAsString());
+
+        if (thingTypeUID.equals(THING_TYPE_SMARTTHINGS)) {
+            bridge = new SmartthingsBridgeHandler((Bridge) thing);
+            registerDeviceDiscoveryService(bridge);
+            return bridge;
+        } else if (SUPPORTED_THING_TYPES_UIDS.contains(thingTypeUID)) {
+            // Everything but the bridge is handled by this one handler
+            logger.debug("Creating thing handler for {}", thingTypeUID.getAsString());
+            SmartthingsThingHandler thingHandler = new SmartthingsThingHandler(thing);
+            thingHandlers.add(thingHandler);
+            return thingHandler;
+        }
+
+        return null;
+    }
+
+    /**
+     * Remove handler of things.
+     */
+    @Override
+    protected synchronized void removeHandler(ThingHandler thingHandler) {
+        if (thingHandler instanceof SmartthingsBridgeHandler) {
+            ServiceRegistration<?> serviceReg = this.discoveryServiceRegs.get(thingHandler.getThing().getUID());
+            if (serviceReg != null) {
+                serviceReg.unregister();
+                discoveryServiceRegs.remove(thingHandler.getThing().getUID());
+            }
+        }
+    }
+
+    /**
+     * Register a new discovery service
+     *
+     * @param handler
+     */
+    private void registerDeviceDiscoveryService(SmartthingsBridgeHandler handler) {
```

Comment:
As said above: Don't start discovery services in the factory. They are either thingHandler services or independent services.